tasks/ThrustersInput.cpp

```diff
@@ -0,0 +1,192 @@
+    	exception(WRONG_SIZE_OF_THRUSTER_COEFFICIENTS);
+    }
+
+
+    // Check size of parameters. They should met among them.
+    if(coeffPos.size() != numberOfThrusters || coeffNeg.size() != numberOfThrusters )
+    {
+    	if(coeffPos.size() != coeffNeg.size())
+    	{	// If coefficients sizes are different among them and controm_modes size, one of the coeff were not set right
+    		RTT::log(RTT::Error) << "Coefficients were not set with "<< numberOfThrusters <<" as predicted" << RTT::endlog();
+    		exception(WRONG_SIZE_OF_THRUSTER_COEFFICIENTS);
+    	}
+    	else if(!controlModes.empty())
+    	{
+    		// Well, if the thruster_coeff vectors have the same size, looks like the control_modes is the one with wrong size
+    		exception(WRONG_SIZE_OF_CONTROLMODES);
+    	}
+    	else if(controlModes.empty())
+    	{
+    		// In this case, control_modes is empty and coeffPos has the right number of thrusters
+    		// Let's set control_modes with default values RAW
+    		numberOfThrusters = coeffPos.size();
+        	//resize the controlModes to number of Thrusters
+        	controlModes.resize(numberOfThrusters);
+        	//set undefined controlModes to RAW
+        	for(unsigned int i = 0; i < numberOfThrusters; i++)
+        	{
+    			controlModes[i] = base::JointState::RAW;
+        	}
+    	}
+    }
+
+    // Check control_modes
+    for (int i=0; i<numberOfThrusters; i++)
+    {
+    	if(controlModes[i] !=  base::JointState::SPEED && controlModes[i] !=  base::JointState::RAW)
+    	{
+			RTT::log(RTT::Error) << "Control Mode should be SPEED or RAW" << RTT::endlog();
+		exception(WRONG_SET_OF_CONTROLMODES);
+    	}
+    }
+
+
+
+
+    return true;
+}
+
+bool ThrustersInput::startHook()
+{
+    if (! ThrustersInputBase::startHook())
+        return false;
+    return true;
+}
+void ThrustersInput::updateHook()
+{
+    ThrustersInputBase::updateHook();
+
+    base::commands::Joints thrusterForces;
+
+    while (_cmd_in.read(thrusterForces) == RTT::NewData)
+    {
+    	if(checkControlInput(thrusterForces))
+		{
+
+    		calcOutput(thrusterForces);
+
+    		_cmd_out.write(thrusterForces);
+		}
+    }
+
+}
+void ThrustersInput::errorHook()
+{
+    ThrustersInputBase::errorHook();
+}
+void ThrustersInput::stopHook()
+{
+    ThrustersInputBase::stopHook();
+}
+void ThrustersInput::cleanupHook()
+{
+    ThrustersInputBase::cleanupHook();
+}
+
+bool ThrustersInput::checkControlInput(base::samples::Joints &cmd_in)
+{
+	std::string textElement;
+
+	// No match of input.size with properties.size
+	if(cmd_in.elements.size() != numberOfThrusters)
+	{
+		RTT::log(RTT::Error) << "The input has not "<< numberOfThrusters <<
+				" thruster as predicted. Actual size is "<< cmd_in.elements.size() << ". Check configuration. "<< RTT::endlog();
+		exception(UNEXPECTED_THRUSTER_INPUT);
+		return false;
+	}
+
+	for (int i=0; i<numberOfThrusters; i++)
+	{
+		// Define how to call the problematic thruster
+		std::string textThruster;
```

Comment:
Don't do this computation if you don't need it, i.e. move it to the if (ERROR) { } below